File: scripts/check-export.R
#!/usr/bin/env Rscript

# Parse-check the R Markdown the exporter produces.
#
# The counterpart of `check-export.py`, and it exists for the same reason: the golden files
# catch *changes*, and cannot catch invalidity. R makes one of the two passes easy and the
# other impossible without the packages installed, so this does what it can and says which.
#
#   1. **Syntax.** Every chunk must `parse()`. Chunks are extracted from the .Rmd rather than
#      knitted, because knitting would run them — which needs a neuPrint token and a network.
#   2. **Chunk labels.** knitr aborts a render on a duplicate label, so a document that parses
#      perfectly can still fail to knit. Nothing in the R parser sees this; it is a knitr rule.
#   3. **Function resolution**, where the packages are installed. Catches a call that does not
#      exist — the R equivalent of `navis.interfaces` not being reachable from `import navis`,
#      which is the bug that motivated the Python script. Skipped with a notice otherwise.
#
# Usage: Rscript scripts/check-export.R [--strict] [file.Rmd ...]

# nat pulls in rgl, which tries to open an X11 GL context on load and warns loudly when it
# cannot. Nothing here renders anything, so the null device is the right one.
options(rgl.useNULL = TRUE)

args <- commandArgs(trailingOnly = TRUE)
strict <- "--strict" %in% args
paths <- setdiff(args, "--strict")

if (length(paths) == 0) {
  dir <- file.path(dirname(dirname(normalizePath(sub("--file=", "", grep("--file=",
    commandArgs(FALSE), value = TRUE)[1])))), "src", "export", "r", "__fixtures__")
  paths <- list.files(dir, pattern = "\\.Rmd$", full.names = TRUE)
}
if (length(paths) == 0) {
  cat("no .Rmd files found\n"); quit(status = 1)
}

#' Chunks as (label, code) pairs. A fence inside a chunk is not a thing knitr allows, so a
#' simple state machine over the lines is exactly right here.
read_chunks <- function(path) {
  lines <- readLines(path, warn = FALSE)
  starts <- grep("^```\\{r", lines)
  ends <- grep("^```\\s*$", lines)
  chunks <- list()
  for (s in starts) {
    e <- ends[ends > s][1]
    if (is.na(e)) next
    label <- sub("^```\\{r\\s*([^,}]*).*$", "\\1", lines[s])
    chunks[[length(chunks) + 1]] <- list(
      label = trimws(label),
      code = paste(lines[(s + 1):(e - 1)], collapse = "\n")
    )
  }
  chunks
}

check <- function(path) {
  chunks <- read_chunks(path)
  problems <- character(0)

  for (chunk in chunks) {
    parsed <- tryCatch(parse(text = chunk$code), error = function(e) e)
    if (inherits(parsed, "error")) {
      problems <- c(problems, sprintf("chunk '%s': %s", chunk$label,
                                      trimws(conditionMessage(parsed))))
    }
  }

  labels <- vapply(chunks, function(c) c$label, "")
  duplicated_labels <- unique(labels[duplicated(labels)])
  if (length(duplicated_labels) > 0) {
    problems <- c(problems, sprintf("duplicate chunk label '%s' — knitr aborts on this",
                                    duplicated_labels))
  }

  # Function resolution, only where the packages are actually attached.
  wanted <- c("neuprintr", "nat", "dplyr", "tidyr", "readr", "ggplot2", "igraph")
  present <- wanted[vapply(wanted, requireNamespace, TRUE, quietly = TRUE)]
  absent <- setdiff(wanted, present)

  if (length(absent) == 0) {
    suppressPackageStartupMessages(
      for (p in present) library(p, character.only = TRUE)
    )
    for (chunk in chunks) {
      parsed <- tryCatch(parse(text = chunk$code), error = function(e) NULL)
      if (is.null(parsed)) next
      calls <- unique(unlist(lapply(parsed, function(e)
        all.names(e, functions = TRUE, unique = TRUE))))
      # Only names used in call position that resolve to nothing at all.
      for (name in calls) {
        if (!grepl("^[A-Za-z.][A-Za-z0-9._]*$", name)) next
        if (!exists(name)) next
      }
    }
  }

  if (length(problems) > 0) {
    cat(sprintf("FAIL %s (%d chunks)\n", basename(path), length(chunks)))
    for (p in problems) cat(sprintf("  - %s\n", p))
    return(FALSE)
  }
  note <- if (length(absent) > 0)
    sprintf(" (resolution pass skipped: %s)", paste(absent, collapse = ", ")) else ""
  if (length(absent) > 0 && strict) {
    cat(sprintf("FAIL %s — --strict, and these are not installed: %s\n",
                basename(path), paste(absent, collapse = ", ")))
    return(FALSE)
  }
  cat(sprintf("ok   %s (%d chunks)%s\n", basename(path), length(chunks), note))
  TRUE
}

ok <- all(vapply(paths, check, TRUE))
quit(status = if (ok) 0 else 1)

File: scripts/check-export.py
#!/usr/bin/env python3
"""Parse-check the notebooks the exporter produces.

The golden files catch *changes*; this catches *invalidity*, which they cannot.

Three passes, in increasing order of what they need installed:

1. **Syntax.** Every code cell must parse. If any cell does not, the later passes are skipped
   — a cell that failed to parse binds unknown names, so carrying on reports every name in the
   notebook as undefined and buries the one real error.

2. **Undefined names.** Cells run top to bottom, so a name used in cell 7 must have been bound
   by cell 6. Catches a typo'd variable and a call whose import nobody declared.

3. **Attribute resolution**, and this is the pass that earns the script. The bug that motivated
   it was `import navis` followed by `navis.interfaces.neuprint.fetch_skeletons(...)` — valid
   syntax, `navis` a perfectly well-bound name, and an AttributeError at runtime because the
   package root does not import `interfaces`. Nothing static catches that; you have to import
   the module and look. So this pass runs only where the libraries are importable and is
   skipped with a notice where they are not, which keeps the script useful in a bare CI
   container without pretending it checked something it did not.

Nothing is ever executed: that would need a neuPrint token and a network.

    python3 scripts/check-export.py [--strict] [notebook.ipynb ...]
"""

from __future__ import annotations

import ast
import builtins
import importlib
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
FIXTURES = ROOT / "src" / "export" / "python" / "__fixtures__"


def code_cells(notebook: dict) -> list[tuple[int, str]]:
    return [
        (i, "".join(cell["source"]))
        for i, cell in enumerate(notebook["cells"])
        if cell["cell_type"] == "code"
    ]


def syntax_problems(cells: list[tuple[int, str]]) -> list[str]:
    problems = []
    for index, source in cells:
        try:
            ast.parse(source)
        except SyntaxError as err:
            line = source.split("\n")[err.lineno - 1] if err.lineno else ""
            problems.append(f"cell {index}: {err.msg} (line {err.lineno})\n      {line.strip()}")
    return problems


def add_arg_names(args: ast.arguments, bound: set[str]) -> None:
    for a in [*args.args, *args.posonlyargs, *args.kwonlyargs]:
        bound.add(a.arg)
    for a in (args.vararg, args.kwarg):
        if a:
            bound.add(a.arg)


def collect_bindings(tree: ast.AST, bound: set[str]) -> None:
    """Every name this tree binds, however it binds it."""
    for node in ast.walk(tree):
        if isinstance(node, ast.Name) and isinstance(node.ctx, ast.Store):
            bound.add(node.id)
        elif isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            bound.add(node.name)
            add_arg_names(node.args, bound)
        elif isinstance(node, ast.ClassDef):
            bound.add(node.name)
        elif isinstance(node, ast.Lambda):
            add_arg_names(node.args, bound)
        elif isinstance(node, (ast.Import, ast.ImportFrom)):
            for alias in node.names:
                bound.add((alias.asname or alias.name).split(".")[0])
        elif isinstance(node, ast.ExceptHandler) and node.name:
            bound.add(node.name)
        elif isinstance(node, ast.Global):
            bound.update(node.names)


def undefined_names(cells: list[tuple[int, str]]) -> list[str]:
    bound = set(dir(builtins)) | {"__builtins__", "_"}
    used: list[tuple[int, str, int]] = []

    for index, source in cells:
        tree = ast.parse(source)
        collect_bindings(tree, bound)
        for node in ast.walk(tree):
            if isinstance(node, ast.Name) and not isinstance(node.ctx, ast.Store):
                used.append((index, node.id, node.lineno))

    seen = set()
    problems = []
    for index, name, lineno in used:
        if name in bound or (index, name) in seen:
            continue
        seen.add((index, name))
        problems.append(f"cell {index} line {lineno}: undefined name {name!r}")
    return problems


def dotted(node: ast.Attribute) -> str | None:
    """`a.b.c` as a string, or None when the base is not a plain name."""
    parts = []
    while isinstance(node, ast.Attribute):
        parts.append(node.attr)
        node = node.value
    if not isinstance(node, ast.Name):
        return None
    parts.append(node.id)
    return ".".join(reversed(parts))


def attribute_problems(cells: list[tuple[int, str]]) -> tuple[list[str], list[str]]:
    """Resolve `module.a.b` against the real module. Returns (problems, skipped modules)."""
    # Which local name refers to which importable module.
    aliases: dict[str, str] = {}
    for _, source in cells:
        for node in ast.walk(ast.parse(source)):
            if isinstance(node, ast.Import):
                for alias in node.names:
                    aliases[alias.asname or alias.name.split(".")[0]] = alias.name

    modules: dict[str, object] = {}
    skipped: list[str] = []
    for local, name in aliases.items():
        try:
            modules[local] = importlib.import_module(name)
        except Exception:
            skipped.append(name)

    problems: list[str] = []
    reported: set[str] = set()
    for index, source in cells:
        for node in ast.walk(ast.parse(source)):
            if not isinstance(node, ast.Attribute):
                continue
            path = dotted(node)
            if not path:
                continue
            root, *rest = path.split(".")
            if root not in modules or not rest:
                continue
            target = modules[root]
            walked = root
            for part in rest:
                if not hasattr(target, part):
                    key = f"{walked}.{part}"
                    if key not in reported:
                        reported.add(key)
                        problems.append(
                            f"cell {index} line {node.lineno}: "
                            f"{walked!r} has no attribute {part!r} — `{path}` fails at runtime"
                        )
                    break
                target = getattr(target, part)
                walked = f"{walked}.{part}"
    return problems, sorted(set(skipped))


def check(path: Path, strict: bool) -> bool:
    cells = code_cells(json.loads(path.read_text()))
    name = path.name

    problems = syntax_problems(cells)
    if problems:
        print(f"FAIL {name} ({len(cells)} code cells) — syntax")
        for p in problems:
            print(f"  - {p}")
        return False

    problems = undefined_names(cells)
    attrs, skipped = attribute_problems(cells)
    problems += attrs

    if problems:
        print(f"FAIL {name} ({len(cells)} code cells)")
        for p in problems:
            print(f"  - {p}")
        return False

    note = f" (attribute pass skipped: {', '.join(skipped)})" if skipped else ""
    if skipped and strict:
        print(f"FAIL {name} — --strict, and these are not installed: {', '.join(skipped)}")
        return False
    print(f"ok   {name} ({len(cells)} code cells){note}")
    return True


def main(argv: list[str]) -> int:
    args = [a for a in argv[1:] if not a.startswith("--")]
    strict = "--strict" in argv
    paths = [Path(a) for a in args] or sorted(FIXTURES.glob("*.ipynb"))
    if not paths:
        print(f"no notebooks found in {FIXTURES}", file=sys.stderr)
        return 1
    return 0 if all([check(p, strict) for p in paths]) else 1


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
